File: practice/algorithms/warmup/diagonal_difference/diagonal_difference.hs
import qualified Data.Matrix as Matrix

diag_sum :: Matrix.Matrix Int -> Int -> Int
diag_sum matrix n = sum [Matrix.getElem k k matrix | k <- [1..n]]

codiag_sum :: Matrix.Matrix Int -> Int -> Int
codiag_sum matrix n = sum [Matrix.getElem (n + 1 - k) k matrix | k <- [1..n]]

main :: IO ()
main = do
    n_temp <- getLine
    m_temp <- getContents
    let n = read n_temp :: Int
        matrix = Matrix.fromLists $ map (map (read :: String -> Int) . words) $ lines m_temp
        diag = diag_sum matrix n
        codiag = codiag_sum matrix n
    print $ abs (diag - codiag)

File: practice/algorithms/warmup/diagonal_difference/diagonal_difference.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

int main()
{
    int n;
    cin >> n;
    vector< vector<int> > matrix(n,vector<int>(n));
    for(int i = 0; i < n; i++)
    {
       for(int j = 0; j < n; j++)
       {
          cin >> matrix[i][j];
       }
    }
    
    int diag = 0;
    int codiag = 0;
    
    for (int i = 0; i < n; i++)
    {
        diag += matrix[i][i];
        codiag += matrix[n - 1 - i][i];
    }
    
    int res = abs(diag - codiag);
    cout << res << endl;
    return 0;
}

